compiler/src/dotty/tools/dotc/ast/tpd.scala

```diff
@@ -1195,6 +1195,13 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
    *    - imports are reflected in the contexts of subsequent statements
    */
   class TreeMapWithPreciseStatContexts(cpy: TreeCopier = tpd.cpy) extends TreeMap(cpy):
+    override def transform(tree: Tree)(using Context): Tree = tree match
+      case Block(stats, expr) =>
+        val stats1 = transformStats(stats :+ expr, ctx.owner)
```

Comment:
This will lead to great tree node churn. Every block will get new statements on every transform, so everything containing a block will be copied, which means you get a complete copy of the AST for a macro transform. I don't think this is acceptable.

Reply:
I think it is better to have an optional `expr` argument in `transformStats`. It's more convoluted but I don't see an alternative.